File: 20-valid-parentheses/20-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    let map = new Map()
    map.set("(", ")")
    map.set("{", "}")
    map.set("[", "]")
    let stack = []
    
    for (let i = 0; i < s.length; i++) {
        let current = s[i]
        if (map.has(current)) {
            stack.push(current)
        } else {
            if (current === map.get(stack[stack.length - 1])) {
                stack.pop()
            } else {
                return false
            }
        }
    }
    let answer
    (stack.length == 0) ? answer = true : answer = false 
    return answer
};

File: 20-valid-parentheses/20-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        HashMap<Character, Character> map = new HashMap<Character, Character>();
        map.put('}', '{');
        map.put(')', '(');
        map.put(']', '[');
        
        Stack<Character> stack = new Stack<Character>();
        
        for (int i = 0; i < s.length(); i++) {
            if (map.containsKey(s.charAt(i))) {
                //remove last element of stack
                char lastElement = stack.isEmpty() ? '#' : stack.pop(); 
                //retrieve pair from map
                char pair = map.get(s.charAt(i));
                //if last element does not equal pair from map
                if (pair != lastElement) {
                    return false;
                }
                //return false
            } else {
                stack.push(s.charAt(i));
            }
        }
        return stack.isEmpty();
    }
}